### DISCUSSION TITLE
Router not working

### DISCUSSION THREAD

**Discussion by devDalys** [vercel/next.js#64675] [App Router]
I have component for logout (http only cookie, not available from document.location, so im using request with Set-cookie)
And router.push('/') does absolutely nothing in the production build, the screenshot shows a variant with router.refresh() which also fails to work
in development mode everything works stably
this isn't the only place I'm encountering this, but there shouldn't be any "buts" as to why it might not work here
![image](https://github.com/vercel/next.js/assets/86551854/305dc9c5-539e-4c46-8459-48f9d883b1d1)
What am I doing wrong? And this behavior is not very stable either
In some random scenario, the transition does happen.
The request is exactly 200, then exactly works.

**Comment by devDalys** [comment]
![image](https://github.com/vercel/next.js/assets/86551854/c459a74a-b90f-4eb5-95ab-766725ad3a18)
and i have this in logs

**Reply by icyJoseph** [reply]
I guess, the problem might be that when you have some logic, during rendering, that redirects to login if there's no cookie, but when you clear the cookie and call router.refresh, you tell the router to:

> A fetch to the server to get the current page in it’s entirety (this includes <html> down to the content, it’s everything on the page but as RSC Payload instead of HTML.

And that is causing some kind of race condition, or just plain wrong behavior.

Would you consider turning your function into a `server action`, defined in another file, and there clear the cookie, because:

> The function calls cookies().set()
> In this case the Router Cache is entirely invalidated and the page is re-rendered, similar to calling revalidateTag, as changing cookies could affect what is rendered in layouts.
> Example: You login / logout
> Example: Feature flags change
> 

See more here, https://github.com/vercel/next.js/discussions/54075
